File: packages/connector-sdk/src/dispatch.ts
import {
  commandRequestSchema,
  replyError,
  replySuccess,
  ERROR_CODES,
  type CommandResponse,
  type ErrorCode,
} from '@cartulaire/connector-contracts'
import { verifyCommandSignature } from '@cartulaire/crypto'

/**
 * En-têtes du transport signé (SPEC §12.1). Dupliqués ici pour que le SDK
 * connecteur ne dépende d'aucun package du cœur.
 */
export const CARTULAIRE_HEADERS = {
  SIGNATURE: 'x-cartulaire-signature',
  TIMESTAMP: 'x-cartulaire-timestamp',
  TRACE_ID: 'x-cartulaire-trace-id',
} as const

/** Contexte fourni à chaque handler de commande. */
export interface CommandContext {
  id: string
  type: string
  traceId: string
  issuer: string
  audience: string
}

/** Handler d'une commande : reçoit le payload validé, renvoie un résultat. */
export type CommandHandler<TPayload = Record<string, unknown>, TResult = unknown> = (
  payload: TPayload,
  ctx: CommandContext,
) => Promise<TResult> | TResult

export interface CommandDefinition {
  type: string
  handler: CommandHandler
}

/** Déclare une commande gérée par le connecteur. */
export function defineCommand<TPayload = Record<string, unknown>, TResult = unknown>(
  type: string,
  handler: CommandHandler<TPayload, TResult>,
): CommandDefinition {
  return { type, handler: handler as CommandHandler }
}

/**
 * Erreur métier levée par un handler pour renvoyer une réponse d'erreur
 * normalisée (ex. INVALID_CREDENTIALS) sans exposer de détail technique.
 */
export class CommandFailure extends Error {
  constructor(
    public readonly code: ErrorCode | string,
    message: string,
    public readonly safeMessage: string,
    public readonly retryable = false,
  ) {
    super(message)
    this.name = 'CommandFailure'
  }
}

export interface DispatchOptions {
  /** Secret HMAC partagé avec l'émetteur (daemon ou cœur). */
  secret: string
  /** Audience attendue par ce connecteur (§13.4). */
  audience: string
  /** Liste blanche des commandes autorisées (§26.4). */
  permissions: readonly string[]
  /** Table des handlers déclarés. */
  handlers: Map<string, CommandHandler>
  /** Fenêtre d'acceptation de l'horodatage (ms). Défaut : 5000. */
  maxSkewMs?: number
  now?: number
}

/**
 * Cœur du connecteur : valide, vérifie la signature et l'audience, applique la
 * liste blanche de permissions, puis dispatche vers le handler. Ne lève jamais —
 * renvoie toujours une réponse normalisée (§13.2/§13.3).
 */
export async function dispatchCommand(
  rawBody: string,
  headers: { signature?: string; timestamp?: string },
  options: DispatchOptions,
): Promise<CommandResponse> {
  const now = options.now ?? Date.now()

  let json: unknown
  try {
    json = JSON.parse(rawBody)
  } catch {
    return replyError('unknown', err(ERROR_CODES.VALIDATION_ERROR, 'Corps JSON invalide'))
  }

  const parsed = commandRequestSchema.safeParse(json)
  if (!parsed.success) {
    return replyError('unknown', err(ERROR_CODES.VALIDATION_ERROR, parsed.error.message))
  }
  const cmd = parsed.data

  // Signature + fraîcheur de l'horodatage
  if (!headers.signature || !headers.timestamp) {
    return replyError(cmd.id, err(ERROR_CODES.INVALID_SIGNATURE, 'En-têtes de signature manquants'))
  }
  const sig = verifyCommandSignature(
    { body: rawBody, timestamp: headers.timestamp, id: cmd.id, type: cmd.type, audience: cmd.audience },
    headers.signature,
    options.secret,
    { maxSkewMs: options.maxSkewMs, now },
  )
  if (!sig.valid) {
    const code =
      sig.reason === 'timestamp_expired' ? ERROR_CODES.COMMAND_EXPIRED : ERROR_CODES.INVALID_SIGNATURE
    return replyError(cmd.id, err(code, `Signature invalide: ${sig.reason}`))
  }

  // Expiration au niveau enveloppe (§13.4)
  if (new Date(cmd.expiresAt).getTime() <= now) {
    return replyError(cmd.id, err(ERROR_CODES.COMMAND_EXPIRED, 'Commande expirée'))
  }

  // Audience exacte (§13.4)
  if (cmd.audience !== options.audience) {
    return replyError(
      cmd.id,
      err(ERROR_CODES.AUDIENCE_MISMATCH, `Audience ${cmd.audience} ≠ ${options.audience}`),
    )
  }

  // Liste blanche de permissions (§26.4)
  if (!options.permissions.includes(cmd.type)) {
    return replyError(cmd.id, err(ERROR_CODES.PERMISSION_DENIED, `Commande ${cmd.type} non autorisée`))
  }

  const handler = options.handlers.get(cmd.type)
  if (!handler) {
    return replyError(cmd.id, err(ERROR_CODES.UNKNOWN_COMMAND, `Commande ${cmd.type} non gérée`))
  }

  try {
    const result = await handler(cmd.payload, {
      id: cmd.id,
      type: cmd.type,
      traceId: cmd.traceId,
      issuer: cmd.issuer,
      audience: cmd.audience,
    })
    return replySuccess(cmd.id, result)
  } catch (e) {
    if (e instanceof CommandFailure) {
      return replyError(cmd.id, {
        code: e.code,
        message: e.message,
        safeMessage: e.safeMessage,
        retryable: e.retryable,
      })
    }
    return replyError(cmd.id, {
      code: ERROR_CODES.INTERNAL_ERROR,
      message: e instanceof Error ? e.message : String(e),
      safeMessage: 'Une erreur est survenue.',
      retryable: false,
    })
  }
}

function err(code: string, message: string) {
  return { code, message, safeMessage: 'Une erreur est survenue.', retryable: false }
}

File: packages/connector-sdk-python/src/cartulaire_connector_sdk/dispatch.py
"""
Cœur du connecteur : valide, vérifie signature/audience/permissions puis dispatche.
Ne lève jamais — renvoie toujours une réponse normalisée (SPEC §13.2/§13.3).
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Dict, Mapping, Optional, Sequence

from .errors import ERROR_CODES, CommandFailure
from .signature import verify_command_signature

# Un handler reçoit (payload, context) et renvoie un résultat sérialisable.
Handler = Callable[[Dict[str, Any], "CommandContext"], Any]


@dataclass(frozen=True)
class CommandContext:
    id: str
    type: str
    trace_id: str
    issuer: str
    audience: str


@dataclass(frozen=True)
class CommandDefinition:
    type: str
    handler: Handler


def define_command(command_type: str, handler: Handler) -> CommandDefinition:
    """Déclare une commande gérée par le connecteur."""
    return CommandDefinition(type=command_type, handler=handler)


def reply_success(command_id: str, result: Any) -> Dict[str, Any]:
    return {"id": command_id, "status": "success", "result": result, "error": None}


def reply_error(
    command_id: str,
    code: str,
    message: str,
    safe_message: str = "Une erreur est survenue.",
    retryable: bool = False,
) -> Dict[str, Any]:
    return {
        "id": command_id,
        "status": "error",
        "result": None,
        "error": {
            "code": code,
            "message": message,
            "safeMessage": safe_message,
            "retryable": retryable,
        },
    }


_REQUIRED_FIELDS = ("id", "type", "issuedAt", "expiresAt", "issuer", "audience", "traceId")


def dispatch_command(
    raw_body: str,
    *,
    signature: Optional[str],
    timestamp: Optional[str],
    secret: str,
    audience: str,
    permissions: Sequence[str],
    handlers: Mapping[str, Handler],
    max_skew_ms: int = 5000,
    now_ms: Optional[int] = None,
) -> Dict[str, Any]:
    """Valide et exécute une commande signée ; renvoie l'enveloppe de réponse."""
    try:
        cmd = json.loads(raw_body)
    except (json.JSONDecodeError, TypeError):
        return reply_error("unknown", ERROR_CODES["VALIDATION_ERROR"], "Corps JSON invalide")

    if not isinstance(cmd, dict) or any(f not in cmd for f in _REQUIRED_FIELDS):
        cmd_id = cmd.get("id", "unknown") if isinstance(cmd, dict) else "unknown"
        return reply_error(cmd_id, ERROR_CODES["VALIDATION_ERROR"], "Enveloppe invalide")

    cmd_id = str(cmd["id"])
    cmd_type = str(cmd["type"])

    # 1) Signature + fraîcheur de l'horodatage
    if not signature or not timestamp:
        return reply_error(cmd_id, ERROR_CODES["INVALID_SIGNATURE"], "En-têtes de signature manquants")

    sig = verify_command_signature(
        body=raw_body,
        timestamp=timestamp,
        command_id=cmd_id,
        command_type=cmd_type,
        audience=str(cmd["audience"]),
        signature=signature,
        secret=secret,
        max_skew_ms=max_skew_ms,
        now_ms=now_ms,
    )
    if not sig.valid:
        code = ERROR_CODES["COMMAND_EXPIRED"] if sig.reason == "timestamp_expired" else ERROR_CODES["INVALID_SIGNATURE"]
        return reply_error(cmd_id, code, f"Signature invalide: {sig.reason}")

    # 2) Expiration au niveau enveloppe (§13.4)
    try:
        exp = datetime.fromisoformat(str(cmd["expiresAt"]).replace("Z", "+00:00"))
        if exp <= datetime.now(timezone.utc):
            return reply_error(cmd_id, ERROR_CODES["COMMAND_EXPIRED"], "Commande expirée")
    except ValueError:
        return reply_error(cmd_id, ERROR_CODES["VALIDATION_ERROR"], "expiresAt invalide")

    # 3) Audience exacte (§13.4)
    if cmd["audience"] != audience:
        return reply_error(cmd_id, ERROR_CODES["AUDIENCE_MISMATCH"], f"Audience {cmd['audience']} ≠ {audience}")

    # 4) Liste blanche de permissions (§26.4)
    if cmd_type not in permissions:
        return reply_error(cmd_id, ERROR_CODES["PERMISSION_DENIED"], f"Commande {cmd_type} non autorisée")

    handler = handlers.get(cmd_type)
    if handler is None:
        return reply_error(cmd_id, ERROR_CODES["UNKNOWN_COMMAND"], f"Commande {cmd_type} non gérée")

    # 5) Exécution
    ctx = CommandContext(
        id=cmd_id,
        type=cmd_type,
        trace_id=str(cmd["traceId"]),
        issuer=str(cmd["issuer"]),
        audience=str(cmd["audience"]),
    )
    try:
        result = handler(cmd.get("payload", {}) or {}, ctx)
        return reply_success(cmd_id, result)
    except CommandFailure as f:
        return reply_error(cmd_id, f.code, f.message, f.safe_message, f.retryable)
    except Exception as e:  # noqa: BLE001
        return reply_error(cmd_id, ERROR_CODES["INTERNAL_ERROR"], str(e))
